listenbrainz/webserver/views/api_tools.py

```diff
@@ -154,12 +154,14 @@ def validate_listen(listen, listen_type):
 
     # Basic metadata
     try:
+        if len(listen['track_metadata']['track_name']) == 0:
+            raise APIBadRequest("required field track_metadata.track_name is empty.", listen)
         if not listen['track_metadata']['track_name']:
-            raise APIBadRequest("JSON document does not contain required "
-                                "track_metadata.track_name.", listen)
```

Comment:
I'm not sure this fix will do exactly what we want. if `['track_name']` exists but is None (null in json) then the `len()` check will fail: `TypeError: object of type 'NoneType' has no len()`. I think for checking an empty value we're still better off with just the original check: `if not listen['track_metadata']['track_name']:`, which will trigger if the item is both None and an empty string (among other things). There is a further test lower down which checks if the artist name is a string... maybe we should also have a check for the track title.

We check for the absense of a track_name or artist_name key implicitly by use of the `except KeyError` at the end of this try block.

Reply:
I reworked the whole section.